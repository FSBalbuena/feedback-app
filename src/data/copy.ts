export const GatherInformation = {
  NEXT_BUTTON: "Next",
  PREV_BUTTON: "Previous",
  GET_FEEDBACK_BUTTON: "Get Feedback",
  TEXT_FIELD_LABEL: "Thoughts...",
};

export const Home = {
  START_BUTTON: "Start",
  TITLE:
    "Confidently give feedback that empowers others and strengthens relationships.",
  DESCRIPTION:
    "If you’re feeling unsure, anxious, or stuck about delivering constructive feedback, this is the place to find clarity, tools, and support. Together, we’ll turn your fears into confidence and your feedback into a positive force for growth.",
};
