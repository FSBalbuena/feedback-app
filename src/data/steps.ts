export const steps = [
  {
    id: "context",
    question:
      "What’s the situation or behavior you’d like to give feedback on?",
    hint: "Is this about a work project, a personal relationship, or something else?",
  },
  {
    id: "perspective",
    question:
      "What specifically did the person do or say that stood out to you?",
    hint: "Describe the actions or behaviors you noticed. Were they positive, negative, or mixed?",
  },
  {
    id: "impact",
    question: "How did this behavior affect you, the team, or the project?",
    hint: "Did it help or hinder progress? How did it make you feel?",
  },
  {
    id: "improvement",
    question: "What changes or improvements would you like to see?",
    hint: "Be specific—what could they do differently next time?",
  },
  {
    id: "strengths",
    question: "What did they do well that you’d like to acknowledge?",
    hint: "Even if there are areas for improvement, what positives can you highlight?",
  },
  {
    id: "tone",
    question: "How would you like to deliver this feedback?",
    hint: "Do you want it to be supportive, direct, or collaborative? Are there any sensitivities to keep in mind?",
  },
  {
    id: "outcome",
    question: "What do you hope to achieve with this feedback?",
    hint: "Is your goal to improve performance, strengthen communication, or something else?",
  },
];

export const stepsIDs = steps.map((step) => step.id);

export const questionsAndHints = steps.reduce((text, step) => {
  const { id, question, hint } = step;
  return text + `- ${id}: ${question} --> ${hint}\n`;
}, "");
