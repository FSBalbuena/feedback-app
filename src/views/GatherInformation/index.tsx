import Grid from "@mui/material/Grid2";
import { useCallback, useState } from "react";
import TextField from "@mui/material/TextField";
import Tooltip from "@mui/material/Tooltip";
import IconButton from "@mui/material/IconButton";
import ArrowBackIos from "@mui/icons-material/ArrowBackIos";
import ArrowForwardIos from "@mui/icons-material/ArrowForwardIos";
import GetAppIcon from "@mui/icons-material/GetApp";
import Typography from "@mui/material/Typography";
import { steps } from "@data/steps";
import { GatherInformation as copyText } from "@data/copy";

type Answers = { [key: string]: string };

type Props = {
  onGetFeedback: (answers: Answers) => void;
};
function GatherInformation({ onGetFeedback }: Props) {
  const [currentStep, setCurrentStep] = useState(0);
  const [answers, setAnswers] = useState<Answers>({});
  const [currentAnswer, setCurrentAnswer] = useState("");

  const isLastStep = currentStep === steps.length - 1;
  const isFirstStep = currentStep === 0;

  const saveAnswer = useCallback(() => {
    const answerId = steps[currentStep].id;
    setAnswers((prevAnswers) => ({
      ...prevAnswers,
      [answerId]: currentAnswer,
    }));
  }, [currentStep, currentAnswer]);

  const onAnswerChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setCurrentAnswer(event.target.value);
    },
    []
  );

  const onMoveToAnswer = useCallback(
    (nextStepIndex: number) => {
      const moveToAnswer = answers[steps[nextStepIndex]?.id];
      saveAnswer();
      setCurrentAnswer(moveToAnswer || "");
    },
    [saveAnswer, answers]
  );

  //--- Buttons Actions ---//
  const onNext = useCallback(() => {
    onMoveToAnswer(currentStep + 1);
    setCurrentStep((previousCurrentStep) => previousCurrentStep + 1);
  }, [onMoveToAnswer, currentStep]);

  const onPrev = useCallback(() => {
    onMoveToAnswer(currentStep - 1);
    setCurrentStep((previousCurrentStep) => previousCurrentStep - 1);
  }, [onMoveToAnswer, currentStep]);

  const handleFeedbackSubmit = () => {
    const answerId = steps[currentStep].id;
    const newAnswers = { ...answers, [answerId]: currentAnswer };
    setAnswers(newAnswers);
    onGetFeedback(newAnswers);
  };
  //--- END Buttons Actions ---//

  return (
    <Grid container spacing={2} maxWidth={800} padding={4}>
      <Grid size={12}>
        <Typography variant={"h4"} gutterBottom>
          {steps[currentStep].question}
        </Typography>
        <Typography variant={"body1"} component={"p"} gutterBottom>
          {steps[currentStep].hint}
        </Typography>
      </Grid>
      <Grid size={12}>
        <TextField
          label={copyText.TEXT_FIELD_LABEL}
          value={currentAnswer}
          onChange={onAnswerChange}
          multiline
          fullWidth
          minRows={4}
          maxRows={6}
        />
      </Grid>
      <Grid size={12} container justifyContent={"space-around"}>
        <Tooltip title={copyText.PREV_BUTTON}>
          <IconButton disabled={isFirstStep} onClick={onPrev}>
            <ArrowBackIos />
          </IconButton>
        </Tooltip>

        {isLastStep ? (
          <Tooltip title={copyText.GET_FEEDBACK_BUTTON}>
            <IconButton
              disabled={!currentAnswer}
              onClick={handleFeedbackSubmit}
            >
              <GetAppIcon />
            </IconButton>
          </Tooltip>
        ) : (
          <Tooltip title={copyText.NEXT_BUTTON}>
            <IconButton disabled={!currentAnswer} onClick={onNext}>
              <ArrowForwardIos />
            </IconButton>
          </Tooltip>
        )}
      </Grid>
    </Grid>
  );
}

export default GatherInformation;
