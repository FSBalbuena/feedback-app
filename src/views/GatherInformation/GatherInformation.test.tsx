import GatherInformation from ".";
import { GatherInformation as copyText } from "../../data/copy";
import { steps } from "../../data/steps";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

const writeAnswer = (answer: string) => {
  const textArea = screen.getByRole("textbox", {
    name: copyText.TEXT_FIELD_LABEL,
  });
  expect(textArea).toBeInTheDocument();
  fireEvent.change(textArea as Element, {
    target: { value: answer },
  });
  waitFor(() => {
    expect(textArea).toHaveValue(answer);
  });
};

const clickOnButton = (buttonIdentifier: string) => {
  const button = screen.getByRole("button", { name: buttonIdentifier });
  expect(button).toBeInTheDocument();
  fireEvent.click(button as Element);
};

const completeOneQuestion = (answer: string) => {
  writeAnswer(answer);
  clickOnButton(copyText.NEXT_BUTTON);
};

const moveToLastQuestion = () => {
  steps.forEach((step, index) => {
    if (index === steps.length - 1) return;
    completeOneQuestion(step.id);
    waitFor(() => {
      const textArea = screen.getByRole("textbox", {
        name: copyText.TEXT_FIELD_LABEL,
      });
      expect(textArea).toHaveValue("");
    });
  });
};

describe("GatherInformation", () => {
  describe("on Render", () => {
    it("should render", () => {
      const { container } = render(<GatherInformation />);
      expect(container).toMatchSnapshot();
    });
  });
  describe("behavior", () => {
    describe("when user is in the first Step", () => {
      it("should show prev button as disabled", () => {
        render(<GatherInformation />);
        const button = screen.getByRole("button", {
          name: copyText.PREV_BUTTON,
        });
        expect(button).toBeInTheDocument();
        expect(button).toBeDisabled();
      });
    });
    describe("when user has not written an answer", () => {
      it("should show next button as disabled", () => {
        render(<GatherInformation />);
        const textArea = screen.getByRole("textbox", {
          name: copyText.TEXT_FIELD_LABEL,
        });
        expect(textArea).toBeInTheDocument();
        const button = screen.getByRole("button", {
          name: copyText.NEXT_BUTTON,
        });
        expect(button).toBeInTheDocument();
        expect(textArea).toHaveValue("");
        expect(button).toBeDisabled();
      });
    });
    describe("when user is not in the first Step", () => {
      it("should show prev button as enabled", () => {
        render(<GatherInformation />);
        const firstAnswer = "first answer";
        writeAnswer(firstAnswer);
        clickOnButton(copyText.NEXT_BUTTON);
        const button = screen.getByRole("button", {
          name: copyText.PREV_BUTTON,
        });
        waitFor(() => {
          expect(button).toBeInTheDocument();
          expect(button).not.toBeDisabled();
        });
      });
    });
    describe("when user is in the last Step", () => {
      it("should not show next button", () => {
        render(<GatherInformation />);
        moveToLastQuestion();
        const button = screen.queryByRole("button", {
          name: copyText.NEXT_BUTTON,
        });
        expect(button).toBeNull();
      });
      it("should show get feedback button  as disabled", () => {
        render(<GatherInformation />);
        moveToLastQuestion();
        const button = screen.getByRole("button", {
          name: copyText.GET_FEEDBACK_BUTTON,
        });
        expect(button).toBeInTheDocument();
        expect(button).toBeDisabled();
      });
      describe("when user has written an answer", () => {
        it("should show get feedback button enabled", () => {
          render(<GatherInformation />);
          moveToLastQuestion();
          writeAnswer("test answer");
          const button = screen.getByRole("button", {
            name: copyText.GET_FEEDBACK_BUTTON,
          });
          expect(button).not.toBeDisabled();
        });
      });
    });
  });

  describe("Interactive", () => {
    describe("when user types on text field", () => {
      it("should update answer", () => {
        const testAnswer = "Test Answer";
        render(<GatherInformation />);
        writeAnswer(testAnswer);
      });
    });
    describe("when user clicks on next button", () => {
      it("should update the title", () => {
        render(<GatherInformation />);
        const title = screen.getByRole("heading");
        expect(title).toBeInTheDocument();
        expect(title).toHaveTextContent(steps[0].question);
        writeAnswer("something");
        clickOnButton(copyText.NEXT_BUTTON);
        waitFor(() => {
          expect(title).toHaveTextContent(steps[1].question);
        });
      });
      it("should update the description", () => {
        render(<GatherInformation />);
        const description = screen.getByRole("paragraph");
        expect(description).toBeInTheDocument();
        expect(description).toHaveTextContent(steps[0].hint);
        writeAnswer("something");
        clickOnButton(copyText.NEXT_BUTTON);
        waitFor(() => {
          expect(description).toHaveTextContent(steps[1].hint);
        });
      });
      it("should update the answer", () => {
        const testAnswer = "Test Answer";
        render(<GatherInformation />);
        writeAnswer(testAnswer);
        clickOnButton(copyText.NEXT_BUTTON);
        waitFor(() => {
          const textArea = screen.getByRole("textbox", {
            name: copyText.TEXT_FIELD_LABEL,
          });
          expect(textArea).toHaveValue("");
        });
      });
    });
    describe("when user clicks on prev field", () => {
      it("should update the title", () => {
        render(<GatherInformation />);
        const title = screen.getByRole("heading");
        expect(title).toHaveTextContent(steps[0].question);
        completeOneQuestion("complete first");
        waitFor(() => {
          expect(title).toHaveTextContent(steps[1].question);
        });

        clickOnButton(copyText.PREV_BUTTON);
        waitFor(() => {
          expect(title).toHaveTextContent(steps[0].question);
        });
      });
      it("should update the description", () => {
        render(<GatherInformation />);
        const description = screen.getByRole("paragraph");
        expect(description).toHaveTextContent(steps[0].hint);
        completeOneQuestion("complete first");
        waitFor(() => {
          expect(description).toHaveTextContent(steps[1].hint);
        });
        clickOnButton(copyText.PREV_BUTTON);
        waitFor(() => {
          expect(description).toHaveTextContent(steps[0].hint);
        });
      });
      it("should update with the previous answer", () => {
        render(<GatherInformation />);
        const firstAnswer = "complete first";
        const textArea = screen.getByRole("textbox", {
          name: copyText.TEXT_FIELD_LABEL,
        });
        completeOneQuestion(firstAnswer);
        waitFor(() => {
          expect(textArea).toHaveValue("");
        });
        clickOnButton(copyText.PREV_BUTTON);
        waitFor(() => {
          expect(textArea).toHaveValue(firstAnswer);
        });
      });
    });
    describe("when user clicks on get Feedback", () => {
      it("should call service", () => {
        render(<GatherInformation />);
        moveToLastQuestion();
        writeAnswer(steps[steps.length - 1].id);
        clickOnButton(copyText.GET_FEEDBACK_BUTTON);
        //actual action is missing
      });
    });
  });
});
