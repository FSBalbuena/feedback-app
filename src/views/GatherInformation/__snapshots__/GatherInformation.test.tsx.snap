// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GatherInformation on Render should render 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-naf2yu-MuiGrid2-root"
  >
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
      >
        What’s the situation or behavior you’d like to give feedback on?
      </h4>
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-vjrxpc-MuiTypography-root"
      >
        Is this about a work project, a personal relationship, or something else?
      </p>
    </div>
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Thoughts...
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-zsp0bh-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-81qg8w"
            >
              <span>
                Thoughts...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-n3xlqn-MuiGrid2-root"
    >
      <button
        aria-label="Previous"
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArrowBackIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
      </button>
      <button
        aria-label="Next"
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ArrowForwardIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
